### ISSUE TITLE
Processes keep running after service is deleted

### ISSUE THREAD

**Issue by guibirow** [Microsoft/service-fabric#257]
I have noticed an odd behavior with service fabric services and this issue is not new, I have seen this before and couldn't understand why, today I found what looks like the reason.

A while ago, I had a few services that used to get messages from a queue, and process them, each process would handle many messages in parallel. One day we had an issue with one of the dependencies of this service and we had to stop them to avoid consuming these messages and drop it to dead letter. After the service were stopped, SF had removed these services from the list of running services, but noticed an odd behavior were the messages in the queue were consumed in the same rate as when the services were running.

We did many checks on cancellation tokens and all possible flaws we cold have in the code to solve it but we never got to a solution.

Recently I was playing around a proof of concept project where I would have a dynamic service being created and destroyed on demand according to the load we received. The idea is to have many different job types, each of them on their own types, instead of sharing a service. To make a good use of the resources, we would remove spare workers for jobs with low demand.

I've noticed, that when we trigger a command to remove a service, event though SF says it has been removed, it disappears from the SF Explorer, but checking on Task Manager, is possible to see these services there for many minutes. Generally between 2 and 5 minutes.

To confirm this was not an issue with the code, I just created a raw Stateless service without any complex logic, just this:

    internal sealed class Worker : StatelessService
    {
        public Worker(StatelessServiceContext context) : base(context) { }

        protected override IEnumerable<ServiceInstanceListener> CreateServiceInstanceListeners()
        {
            return this.CreateServiceRemotingInstanceListeners();
        }

        protected override async Task RunAsync(CancellationToken cancellationToken)
        {
            while (true)
            {
                cancellationToken.ThrowIfCancellationRequested();
                await Task.Delay(10000, cancellationToken);
            }
        }

With this I could still notice the same behavior, and this make impractical to have dynamic services with a threshold lower that 5min without wasting resources.

PS: I know I can design the services in different ways to have similar behavior, but I want to understand in this scenarios the following:

- If the service didn't actually stopped, why SF does not show them in the explorer?
- Why they have to hang for so long to shutdown?

Regards,
Diego

**Comment by VipulM-MSFT** [comment]
@guibirow - The replica lifecycle and process lifecycle are different. When you call delete service, the replicas associated with the services are closed and removed from the process. Once a last replica from the process is removed, the process is closed after a configurable grace interval. If during the grace interval a new replica is placed in the process because of load balancing, failover or creation of new partition/service, the system can avoid "activation" phase, which involves setting up the folder, performing necessary ACLs, creating process etc. The grace interval is configurable through the [DeactivationGraceInterval in Hosting](https://github.com/Microsoft/service-fabric/blob/master/src/prod/src/Hosting2/HostingConfig.h) section of the fabric settings.

**Comment by guibirow** [comment]
>  When you call delete service, the replicas associated with the services are closed and removed from the process. Once a last replica from the process is removed, the process is closed after a configurable grace interval.

Thanks @VipulM-MSFT, it makes things clear now. 

I don't remember exactly, but If I am not wrong, when we had the process consuming messages from the queue, I think we did create ThreadPool threads and left them running, assuming a service deletion would kill the process with all threads, if the process keep running the threads would still be running in the ThreadPool, this explain everything. 

If I understood correctly, SF will scan for deactivation (**_DeactivationScanInterval_**) every 10min (_default_), when hosting mode is set to shared process, if any replica is not allocated in the last 1min(_DeactivationGraceInterval_) to the same process, it will be marked for deactivation and then deleted. AM I correct?

If yes, this clearly explain why some processes where left behind.
When the hosting mode is Exclusive, the deactivation(_ExclusiveModeDeactivationGraceInterval_) should happen in 1sec (default), but because the default scan interval is 10 min it will take much longer, in this case the process will be around until the next scan happens? am I correct?

My questions now are:

Does the kill process happens after the scan in a later phase or the scan also kills the process?

Are these settings configured on a cluster basis or can we set then per service type or application?

What are the odds of setting these scan intervals to lower intervals like: 30sec or 1min? Mainly for _DeactivationScanInterval_

**Comment by guibirow** [comment]
OK, for shared process is clear to me now.

I am still not understanding why ExclusiveProcess keeps around for a while, based on the defaults they should be killed almost instantaneously after deactivation. If **_DeactivationScanInterval_** is not responsible for removing them, what might causing this delay? is there any other configuration I didn't notice?

**Comment by biltongza** [comment]
I see this issue all the time, it's annoying when debugging and changing code because it blocks the publish process from visual studio, it says the executable is locked and times out after 10 retries. I can remove the services manually via SF explorer, kill the processes in task manager and watch them respawn a few seconds later. Only way to get around it is to restart the cluster. It's an annoyance that slows down development.
